File: src/quest_editor/model/WaveModel.ts
import { Property } from "../../core/observable/property/Property";
import { WritableProperty } from "../../core/observable/property/WritableProperty";
import { require_non_negative_integer } from "../../core/util";
import { property } from "../../core/observable";

export class WaveModel {
    private readonly _id: WritableProperty<number>;
    private readonly _area_id: Property<number>;
    private readonly _section_id: Property<number>;

    readonly id: Property<number>;
    readonly area_id: Property<number>;
    readonly section_id: Property<number>;

    constructor(id: number, area_id: number, section_id: number) {
        require_non_negative_integer(id, "id");
        require_non_negative_integer(area_id, "area_id");
        require_non_negative_integer(section_id, "section_id");

        this._id = property(id);
        this.id = this._id;

        this._area_id = property(area_id);
        this.area_id = this._area_id;

        this._section_id = property(section_id);
        this.section_id = this._section_id;
    }
}

File: web/src/main/kotlin/world/phantasmal/web/questEditor/models/WaveModel.kt
package world.phantasmal.web.questEditor.models

import world.phantasmal.observable.value.Val
import world.phantasmal.observable.value.mutableVal

class WaveModel(id: Int, areaId: Int, sectionId: Int) {
    private val _id = mutableVal(id)
    private val _areaId = mutableVal(areaId)
    private val _sectionId = mutableVal(sectionId)

    val id: Val<Int> = _id
    val areaId: Val<Int> = _areaId
    val sectionId: Val<Int> = _sectionId
}
